### PR TITLE
Add screen navigation methods for Speculos backend.

### PR DESCRIPTION
**Pull request by agrojean-ledger**

### CODE REVIEW

File: src/ragger/backend/speculos.py [hunk 1]
```diff
@@ -15,9 +15,12 @@
 """
 from pathlib import Path
 from contextlib import contextmanager
+from io import BytesIO
+from time import sleep, time
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/speculos.py:19]: I removed the unused import.

File: tests/stubs.py [hunk 2]
```diff
@@ -26,6 +30,26 @@ def apdu(*args):
...
 def button(*args):
     return {}
+iter = 0
+idx = 1
+def screenshot(*args):
+    global iter
+    global idx
+    path = Path(__file__).parent.resolve() / "snapshots/nanos/generic" / f"{str(idx).zfill(5)}.png" 
+    sleep(0.25)
+    if iter==1:
```
Comment (agrojean-ledger) [inline tests/stubs.py:40]: @lpascal-ledger ugly but I don't really know how to stub it in a smarter way.

Comment (lpascal-ledger) [reply tests/stubs.py:40]: Maybe with a class? They are useful as local namespaces:
```python
class Screenshots:
    def __init__(self):
        self.iter = 0
        self.idx = 1
        self.steps = {1: 0, 5: 1, 8: 2, 13: 3}

    def fetch(self):
        path = Path(__file__).parent.resolve() / "snapshots/nanos/generic" / f"{str(self.idx).zfill(5)}.png"
        self.idx = self.steps.get(self.iter, self.idx)
        self.iter += 1
        img_temp = Image.open(path)
        iobytes = BytesIO()
        img_temp.save(iobytes, format="PNG")
        return iobytes.getvalue(),200
[...]

class SpeculosServerStub:
    def __init__(self):
        screenshots = Screenshots()
        [...]
        self.app.add_url_rule("/screenshot", methods=["GET"], view_func=screenshots.fetch)
```

(Was the `sleep` necessary?)

Comment (agrojean-ledger) [reply tests/stubs.py]: @lpascal-ledger I updated the archi based on your advice and with a smarter approach (use button requests to update index of image, of course !)

`sleep` was necessary with first implem. I removed it.

File: tests/backend/test_speculos.py [hunk 3]
```diff
@@ -149,3 +149,41 @@ def test_clicks(self):
                     self.backend.right_click()
                     self.backend.left_click()
                     self.backend.both_click()
+    
+    def test_navigate_until_snap(self):
+        with patch("speculos.client.subprocess"):
+            with SpeculosServerStub():
+                with self.backend:
+                    ret = self.backend.navigate_until_snap(Path(__file__).parent.parent.resolve(),"generic",0,2)
```
Comment (lpascal-ledger) [inline tests/backend/test_speculos.py:157]: As it's used several times, you could define a constant with the path prefix on top of the file.
```python
ROOT_SCREENSHOT_PATH = Path(__file__).parent.parent.resolve()
```

File: tests/backend/test_speculos.py [hunk 4]
```diff
@@ -149,3 +149,41 @@ def test_clicks(self):
...
+                self.assertEqual(ret,2)
+    
+    def test_navigate_and_compare_until_snap(self):
+        with patch("speculos.client.subprocess"):
+            with SpeculosServerStub():
+                with self.backend:
+                    ret = self.backend.navigate_and_compare_until_snap(Path(__file__).parent.parent.resolve(),"generic",0,2)
+                self.assertEqual(ret,True)
```
Comment (lpascal-ledger) [inline tests/backend/test_speculos.py:165]: `self.assertTrue(ret)` ?

File: tests/stubs.py [hunk 5]
```diff
@@ -1,10 +1,13 @@
...
 from enum import IntEnum
+from io import BytesIO
 from multiprocessing import Process
+from pathlib import Path
 
 from flask import Flask, request
+from PIL import Image
 from requests import get
 from requests.exceptions import ConnectionError
-
```
Comment (lpascal-ledger) [inline tests/stubs.py:7]: The test did not deserve that extra blank line?

File: tests/stubs.py [hunk 6]
```diff
@@ -24,22 +27,49 @@ def apdu(*args):
...
+        self.idx = 1
+    
+    def button(self,*args):
+        path = request.path
+        if "right" in path:
+            self.idx = self.idx+1 if (self.idx < 2) else 0
+        elif "left" in path:
+            self.idx = self.idx-1 if (self.idx >= 1) else 0
+        elif "both" in path:
+            if self.idx==2:
```
Comment (lpascal-ledger) [inline tests/stubs.py:43]: There are a lot of inconsistency with spaces before/after operators, comas, etc. Maybe we should add `yapf` on tests too.

**Review by lpascal-ledger — APPROVED** [review]
Nothing more to say than some non-blocking cosmetic ranting.

File: tests/backend/test_speculos.py (hunk 4 above)
Comment (agrojean-ledger) [reply tests/backend/test_speculos.py:165]: Done

File: tests/backend/test_speculos.py (hunk 3 above)
Comment (agrojean-ledger) [reply tests/backend/test_speculos.py:157]: Done.

File: tests/stubs.py (hunk 6 above)
Comment (agrojean-ledger) [reply tests/stubs.py]: Done.

File: tests/stubs.py (hunk 5 above)
Comment (agrojean-ledger) [reply tests/stubs.py:7]: Ran yapf, line is back.

File: src/ragger/backend/speculos.py [hunk 7]
```diff
@@ -119,3 +122,139 @@ def left_click(self) -> None:
...
+                "left": 0,
+                "upper": 0,
+                "right": 0,
+                "lower": 0
+            }) -> int:
+
+        snapshot_golden_path = path / "snapshots" / self._firmware.device / test_case_name
+        snapshot_tmp_path = path / "snapshots-tmp" / self._firmware.device / test_case_name
+
+        snapshot_golden_path.mkdir(parents=True, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/speculos.py:188]: @fbeutin-ledger, actually `exist_ok` raises an `FileExistsError` if the directory **already exists** when set to `False`.
But I'll add an error if the golden folder does not exist.

[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/speculos.py:194]: Done @fbeutin-ledger

File: src/ragger/backend/ledgercomm.py [hunk 8]
```diff
@@ -100,3 +101,32 @@ def left_click(self) -> None:
...
+                "lower": 0
+            }) -> int:
+        pass
+
+    def navigate_and_compare_until_snap(self,
+                                        path: Path,
+                                        test_case_name: Path,
+                                        start_img_idx: int = 0,
+                                        last_img_idx: int = 0) -> bool:
+        pass
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/ledgercomm.py:132]: Ok

File: src/ragger/backend/ledgercomm.py [hunk 9]
```diff
@@ -100,3 +101,32 @@ def left_click(self) -> None:
...
+                "right": 0,
+                "lower": 0
+            },
+            crop_last: dict = {
+                "left": 0,
+                "upper": 0,
+                "right": 0,
+                "lower": 0
+            }) -> int:
+        pass
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/ledgercomm.py:125]: Ok, in new issue/PR ?

File: src/ragger/backend/interface.py [hunk 10]
```diff
@@ -286,3 +286,87 @@ def both_click(self) -> None:
...
+                "left": 0,
+                "upper": 0,
+                "right": 0,
+                "lower": 0
+            },
+            crop_last: dict = {
+                "left": 0,
+                "upper": 0,
+                "right": 0,
+                "lower": 0
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py:340]: @fbeutin-ledger if I had the ` Optional[(dict, dict)]` it goes against the idea of guiding the user, doesn't it ? It wouldn't be explicit what the dict should be

File: src/ragger/backend/ledgerwallet.py [hunk 11]
```diff
@@ -99,3 +100,32 @@ def left_click(self) -> None:
...
+                "upper": 0,
+                "right": 0,
+                "lower": 0
+            },
+            crop_last: dict = {
+                "left": 0,
+                "upper": 0,
+                "right": 0,
+                "lower": 0
+            }) -> int:
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/ledgerwallet.py:123]: See comment https://github.com/LedgerHQ/ragger/pull/31/files#r996764978

File: src/ragger/backend/ledgercomm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/ragger/backend/ledgercomm.py:125]: Ok with this

File: src/ragger/backend/interface.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (fbeutin-ledger) [reply src/ragger/backend/interface.py:340]: True, maybe a Optional[(CropDirective, CropDirective)]
With
class CropDirective:
     __init__(self, etc....

**Review by fbeutin-ledger — APPROVED** [review]

File: src/ragger/backend/interface.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ragger/backend/interface.py]: Done.

**Review by fbeutin-ledger — APPROVED** [review]

**Review by lpascal-ledger — DISMISSED** [review]

**Review by lpascal-ledger — APPROVED** [review]

**Pull request merged**